src/main/java/org/gaul/s3proxy/EncryptedBlobStore.java

```diff
@@ -0,0 +1,766 @@
+        PageSet<? extends StorageMetadata> pageSet) {
+        ImmutableSet.Builder<StorageMetadata> builder = ImmutableSet.builder();
+        for (StorageMetadata sm : pageSet) {
+            if (sm instanceof BlobMetadata) {
+                MutableBlobMetadata mbm =
+                    new MutableBlobMetadataImpl((BlobMetadata) sm);
+
+                // if blob is encrypted remove the -s3enc suffix
+                // from content type
+                if (isEncrypted(mbm)) {
+                    mbm = removeEncryptedSuffix((BlobMetadata) sm);
+                    mbm = calculateBlobSize(mbm);
+                }
+
+                builder.add(mbm);
+            } else {
+                builder.add(sm);
+            }
+        }
+
+        // make sure the marker do not show blob with .s3enc suffix
+        String marker = pageSet.getNextMarker();
+        if (marker != null && isEncrypted(marker)) {
+            marker = removeEncryptedSuffix(marker);
+        }
+        return new PageSetImpl<>(builder.build(), marker);
+    }
+
+    private boolean isEncrypted(BlobMetadata blobMeta) {
+        return isEncrypted(blobMeta.getName());
+    }
+
+    private boolean isEncrypted(String blobName) {
+        return blobName.endsWith(Constants.S3_ENC_SUFFIX);
+    }
+
+    private MutableBlobMetadata setEncryptedSuffix(BlobMetadata blobMeta) {
+        MutableBlobMetadata bm = new MutableBlobMetadataImpl(blobMeta);
+        if (blobMeta.getName() != null && !isEncrypted(blobMeta.getName())) {
+            bm.setName(blobNameWithSuffix(blobMeta.getName()));
+        }
+
+        return bm;
+    }
+
+    private String removeEncryptedSuffix(String blobName) {
+        return blobName.substring(0,
+            blobName.length() - Constants.S3_ENC_SUFFIX.length());
+    }
+
+    private MutableBlobMetadata removeEncryptedSuffix(BlobMetadata blobMeta) {
+        MutableBlobMetadata bm = new MutableBlobMetadataImpl(blobMeta);
+        if (isEncrypted(bm.getName())) {
+            String blobName = bm.getName();
+            bm.setName(removeEncryptedSuffix(blobName));
+        }
+
+        return bm;
+    }
+
+    private MutableBlobMetadata calculateBlobSize(BlobMetadata blobMeta) {
+        MutableBlobMetadata mbm = removeEncryptedSuffix(blobMeta);
+
+        // we are using on non-s3 backends like azure or gcp a metadata key to
+        // calculate the part padding sizes that needs to be removed
+        if (mbm.getUserMetadata()
+            .containsKey(Constants.METADATA_ENCRYPTION_PARTS)) {
+            int parts = Integer.parseInt(
+                mbm.getUserMetadata().get(Constants.METADATA_ENCRYPTION_PARTS));
+            int partPaddingSizes = Constants.PADDING_BLOCK_SIZE * parts;
+            long size = blobMeta.getSize() - partPaddingSizes;
+            mbm.setSize(size);
+            mbm.getContentMetadata().setContentLength(size);
+        } else {
+            // on s3 backends like aws or minio we rely on the eTag suffix
```

Comment:
I think this assumption is problematic when a server side copy is performed for objects < 5GB that were uploaded via MPU, because the new blob will not have the MPU ETag. This shows up when doing a server side copy (i.e. PutBlob with CopySource being the encrypted blob). The way around this is to convert this copy request into a new multipart upload, server-side copy the parts, then complete the upload.

I suggest failing the server side request in this case, at least to start with. That's how multipart upload server side copy has to be performed anyway.

I tested the behavior with a ~6MB object and ended up reading 2560 1024 byte chunks back after the server side copy was performed.

Reply:
I will try to look this week into it

Reply:
this part of the code only affects the file listing functions. If you upload two parts and complete the upload then server side copy the file, it will loose the part count on the eTag but the decryption is still working as it detects the part count from the padding. But maybe it don't really get the issue, if so please help me with some s3cmd or curl commands to reproduce this behavior.